Allabana/cc sysroot creator and mixins

piraka9011 opened the pull request:
Includes the following changes:
- TQDM progress bar with output from docker as postfix.
- Requirements.txt file and .gitignore
- Some documentation updates
- Mixins for armhf with docs

Commit:
Added spinner to docker build image

Commit:
Spinners, Requirements, Docs

Commit:
Install script and doc updates

Commit:
Update to output Docker stdout to tool output

Commit:
Modify argparse code to check for argument validity better

thomas-moulard commented on .gitignore:
```diff
@@ -0,0 +1,13 @@
+qemu-user-static/
+*armhf-*/
+*_armhf-*/
+*_x86-*/
```
nit: newline

thomas-moulard commented on INSTALL_MAC.md:
```diff
@@ -0,0 +1,60 @@
+
+### Install prerequisites
+
+The cross compilation toolchain and docker have to be installed. The following instruction have been tested on Ubuntu Xenial.
```
Isn't it a bit contradictory - those are OS X instructions but tested on Ubuntu?

thomas-moulard commented on INSTALL_MAC.md:
````diff
@@ -0,0 +1,60 @@
+The cross compilation toolchain and docker have to be installed. The following instruction have been tested on Ubuntu Xenial.
+
+```bash
+# Make sure you're ready to install stuff
````
Ensure that your brew install is healthy? So that people understand why we need this.

thomas-moulard commented on INSTALL_MAC.md:
````diff
@@ -0,0 +1,60 @@
+sudo python3 -m pip install -U  colcon-common-extensions rosdep vcstool
+```
+
+Install Docker toolbox [here](https://docs.docker.com/toolbox/toolbox_install_mac/)
````
```suggestion
[Install Docker toolbox](https://docs.docker.com/toolbox/toolbox_install_mac/)
```

thomas-moulard commented on INSTALL_MAC.md:
```diff
@@ -0,0 +1,60 @@
+# Make sure you're ready to install stuff
+brew doctor
+brew install cmake git wget curl bash-completion qemu
+sudo python3 -m pip install -U  colcon-common-extensions rosdep vcstool
```
What about installing this locally to avoid impacting the whole system if we can avoid it? Especially, we install packages locally below so we do it anyway.

```suggestion
python3 -m pip install --user -U  colcon-common-extensions rosdep vcstool
```

thomas-moulard commented on INSTALL_MAC.md:
````diff
@@ -0,0 +1,60 @@
+
+```bash
+# Install plugin
+pip3 install --editable src/Colcon-cc-build/colcon_cc_build --user
````
```suggestion
python3 -m pip install --editable src/Colcon-cc-build/colcon_cc_build --user
```

thomas-moulard commented on INSTALL_MAC.md:
````diff
@@ -0,0 +1,60 @@
+
+```bash
+# Install plugin
+pip3 install --editable src/Colcon-cc-build/colcon_cc_build --user
````
Why editable?

piraka9011 replied:
This is from Juan's docs. Not sure why so will need to ask.

thomas-moulard commented on INSTALL_MAC.md:
```diff
@@ -0,0 +1,60 @@
+## 1. setup the sysroot
+## add --force-sysroot-build to force rebuilding the sysroot
+colcon cc-setup-sysroot --arch generic_armhf --os ubuntu_bionic \
+   --sysroot-base-image juanrh/ros2:armhf_bionic_crystal_fastrtps_prebuilt-crystal
```
`juanrh` ? 😆 
Do we need a plan to get a better place to hold those images?

piraka9011 replied:
AWS ECR maybe?

thomas-moulard commented on INSTALL_MAC.md:
```diff
@@ -0,0 +1,60 @@
+## 2. Complete cc-build setup following the instructions in the output of the previous command
+bash generic_armhf-ubuntu_bionic-fastrtps-crystal/cc_system_setup.bash
+source generic_armhf-ubuntu_bionic-fastrtps-crystal/cc_build_setup.bash
+## 3. launch cross compilation using that sysroot
```
```suggestion
## 3. Launch cross compilation using that sysroot
```

thomas-moulard commented on INSTALL_MAC.md:
```diff
@@ -0,0 +1,60 @@
+pip3 install --editable src/Colcon-cc-build/colcon_cc_build --user
+
+# Launch a build
+## 1. setup the sysroot
```
```suggestion
## 1. Setup the sysroot
```

thomas-moulard commented on INSTALL_MAC.md:
```diff
@@ -0,0 +1,60 @@
+Additional base images can be created using the script `colcon_cc_build/colcon_cc_build/verb/sysroot/publish_to_ECR.sh`.
+Some base images already available to use with `--sysroot-base-image`:
+
+- `juanrh/ros2:armhf_bionic_crystal_fastrtps_prebuilt-crystal` for building ROS 2 packages starting from a prebuilt binary for ROS 2 crystal for armhf.
```
Ditto for the `juanrh`. We probably should figure that out before launch?

thomas-moulard commented on INSTALL_MAC.md:
```diff
@@ -0,0 +1,60 @@
+
+### Debug
+
+Manually build and/or run the workspace image
```
```suggestion
Manually build and/or run the workspace image.
```

thomas-moulard commented on INSTALL_MAC.md:
```diff
@@ -0,0 +1,60 @@
+docker image build -f colcon_cc_build/colcon_cc_build/verb/sysroot/Dockerfile_workspace \
+  --network host \
+  -t ros2_benchmark_pipeline:latest \
+  --build-arg ROS2_BASE_IMG=913674827342.dkr.ecr.us-west-2.amazonaws.com/ros2:ubuntu_arm-crystal \
```
It'd be nice to document why we use this location.

thomas-moulard commented on INSTALL_MAC.md:
```diff
@@ -0,0 +1,60 @@
+  --network host \
+  -t ros2_benchmark_pipeline:latest \
+  --build-arg ROS2_BASE_IMG=913674827342.dkr.ecr.us-west-2.amazonaws.com/ros2:ubuntu_arm-crystal \
+  --build-arg ROS2_WORKSPACE=. --build-arg ROS_DISTRO=crystal --build-arg TARGET_TRIPLE=aarch64-linux-gnu \
```
Could we do dashing instead? That seems more appropriate.

thomas-moulard commented on mixins/README.md:
````diff
@@ -0,0 +1,18 @@
+colcon mixin add cc_mixins file://<path_to_mixin_directory>/index.yaml
+colcon mixin update cc_mixins
+```
+where `cc_mixins` is the name of your
````
your?

thomas-moulard commented on mixins/cross-compile.mixin:
```diff
@@ -0,0 +1,31 @@
+        },
+        "pi_armhf-raspbian_stretch": {
+            "cmake-args": [
+                "-DCMAKE_SYSTEM_NAME=Linux",
```
Why do we pass values here which are statically set in the cmake toolchain file?

prajakta-gokhale replied:
We want to replace the toolchain files with colcon mixins for all supported architectures. Synced with Anas and we will do that in the next iteration.

thomas-moulard commented on mixins/cross-compile.mixin:
```diff
@@ -0,0 +1,31 @@
+                "-DCMAKE_C_COMPILER=/usr/bin/arm-linux-gnueabihf-gcc",
+                "-DCMAKE_CXX_COMPILER=/usr/bin/arm-linux-gnueabihf-g++",
+                "-DCMAKE_LIBRARY_ARCHITECTURE=arm-linux-gnueabihf",
+                "-DCMAKE_FIND_ROOT_PATH=/home/ubuntu/aws-ros-dev/cross_compile/sysroot_creator/scripts/armhf-ubuntu-fastrtps-dashing/sysroot",
```
Not portable?

thomas-moulard commented on mixins/install.sh:
```diff
@@ -0,0 +1,9 @@
+  colcon mixin update cc_mixins
+else
+  echo "Linting failed. Check your linting and run again to install mixins."
+fi
```
nit: newline

thomas-moulard commented on mixins/lint.py:
```diff
@@ -0,0 +1,38 @@
+            name,
+        ])
+    except SystemExit as e:
+        any_error |= bool(e.code)
```
Shouldn't we log something here?

thomas-moulard commented on mixins/lint.py:
```diff
@@ -0,0 +1,38 @@
+    except SystemExit as e:
+        any_error |= bool(e.code)
+        continue
+    assert False, 'run() should always raise SystemExit'
```
Need a comment there - why?

piraka9011 replied:
This was taken from the [`colcon-mixin-repository`](https://github.com/colcon/colcon-mixin-repository).
Updated docstrings to reflect accordingly and modified a bit to log/update.

thomas-moulard commented on sysroot_creator/scripts/create_cc_sysroot:
```diff
@@ -81,10 +83,11 @@ class DockerConfig:
         self.base_image = base_image
         self.network_mode = network_mode
         self.nocache = nocache
+        self.ros2_workspace = ros2_workspace
```
If this is not meant to be exposed, it'd be better to make this protected. (`self._ros2_workspace`)

thomas-moulard commented on sysroot_creator/scripts/create_cc_sysroot:
```diff
@@ -143,20 +146,29 @@ def build_workspace_sysroot_image(
+                        continue
+                    t.set_postfix_str(line)
+                except (KeyError, AttributeError) as e:
+                    print("Error reading the Dockerfile image. The following exception"
```
Better use a real logging library when possible.
print => logging.exception

thomas-moulard commented on sysroot_creator/scripts/create_cc_sysroot:
```diff
@@ -143,20 +146,29 @@ def build_workspace_sysroot_image(
+                        continue
+                    t.set_postfix_str(line)
+                except (KeyError, AttributeError) as e:
+                    print("Error reading the Dockerfile image. The following exception"
```
Is this reading the `Dockerfile` image?

piraka9011 replied:
Yes, its also yielding the logs from the build output.

thomas-moulard commented on sysroot_creator/scripts/create_cc_sysroot:
```diff
@@ -222,67 +234,84 @@ def write_cc_system_setup_script(cc_root_dir: Path, platform: Platform):
         type=str,
-        help='Base image to use for building the sysroot')
+        help='Base Docker image to use for building the sysroot.'
+             'Ex. juanrh/ros2:armhf_bionic_crystal_fastrtps_prebuilt-crystal')
```
`juanrh`?

thomas-moulard commented on sysroot_creator/scripts/create_cc_sysroot:
```diff
@@ -222,67 +234,84 @@ def write_cc_system_setup_script(cc_root_dir: Path, platform: Platform):
+        '--ros2-workspace',
+        required=False,
+        type=str,
+        default='~/ros2_ws',
```
The `~` won't be expended automatically. Do we handle this on our side?

thomas-moulard reviewed (commented):
Generally LGTM.

A few comments:
- We should avoid committing un-typed Python3 code.
- This PR is too big. This should at least have been one PR for the doc, one PR for the arg fix, one PR for the Dockerfile.

Let's split responsiblities properly from next time. As PR commits are combined when the PR gets merged, basically it means we can't rollback any commit on master for this project which is a problem.

Commit:
Update for CR:
- Updated mixins to reflect toolchain file
- Address nits
- Fix docs
- Updated requirements
- Switch print to logging
TODO: Update default Dockerfile args.

piraka9011 commented:
Updated with requested changes.
Regarding the Dockerfiles: We've made an AWS ECR Repo to which we'll upload the Dockerfiles we want to use. 
Will update soon.

Commit:
- Automatically setup build environment at the end of script
- Add progress bar for main pipeline

prajakta-gokhale commented on sysroot_creator/scripts/create_cc_sysroot:
```diff
@@ -277,33 +320,50 @@ def main():
+    pbar, step = update_pbar(pbar, step)
+    pbar.close()
+
+    logger.warning("""
```
This can be removed now, since it is done from the script itself now?

piraka9011 replied:
Left it so users could know in the future how to setup their environment. I think it might be better to combine them into one script tho.

prajakta-gokhale commented on INSTALL_MAC.md:
```diff
@@ -0,0 +1,61 @@
+
+[Install Docker toolbox](https://docs.docker.com/toolbox/toolbox_install_mac/)
+
+### Building a workspace
```
With the new implementation approach, most of these steps will be invalid. I think it's better to remove them from here if you're planning to move forward with this PR.

The Mac installation steps should only contain the Mac-specific part and the rest should stay in the main README.

prajakta-gokhale reviewed (commented):
As discussed, let's split this PR into smaller PRs:
1. tqdm and progress bar changes
2. README, Installation changes
3. Colcon mixins

That way we can get them merged faster and make it easier to review and test.

Commit:
Update README

Commit:
Update Dockerfile defaults

prajakta-gokhale commented on README.md:
```diff
@@ -2,11 +2,12 @@
 
-The cross compilation toolchain and docker have to be installed. The following instruction have been tested on Ubuntu Xenial.
+The cross compilation toolchain and docker have to be installed. 
+The following instruction have been tested on Ubuntu Xenial (18.04).
```
Nit: [Xenial](http://releases.ubuntu.com/16.04/) is `16.04`. `18.04` is [Bionic](http://releases.ubuntu.com/18.04/) :)

piraka9011 replied:
Oops... That shouldn't have slipped...

Commit:
Change logger to info

piraka9011 committed:
Merge branch 'cc-sysroot-creator' into allabana/cc-sysroot-creator

piraka9011 commented:
@thomas-moulard @prajakta-gokhale Please take a look at the following changes.

prajakta-gokhale reviewed (approved):
Looks good to me. 

The missing pieces in the README and installation will be filled up once we merge all our changes.

Commit:
Update README.md
